### PR TITLE
Add target SPRACINGF7DUAL

### PR DESCRIPTION
**Pull request by sdirkwinkel**
This pull request imports the SPRACINGF7DUAL from Betaflight and gets it building. I've separated importing and fixing into different commits for better accountability.

This currently does not work for me: After flashing my board doesn't get a serial connection in the configurator. I'll propably get a debugger at some point to see what is happening. The board works fine with Cleanflight.

I'd appreciate any pointers as to where to start looking for mistakes as I'm new to inav(and cf/bf).

Ping @hydra: consider removing inav from your description until this works

### CODE REVIEW

**Comment by hydra** [comment]
Hiya, there are some things to note regarding the SPRacingF7DUAL.

1) It has two gyros, both gyro SPI cable select pins MUST be initialized and set to inactive before attempting to use either of the gyros. Certainly you want to check the hardware initialisation of the gyro pins as the first thing to do otherwise it will NOT be reliable to a) fly, b) initialise consistently - especially on cold boots.  See the CF/BF gyro dual initialisation code for reference.
2) UART5 RX is physically shared with the BIND button pin.  Ensure there is no conflict between button and uart code.
3) Motor outputs 5-8 are used to control the FX758-2-SPI VTX, when not using the VTX they are free for motor outputs, ensure any motor and vtx code is not conflicting.


I've never worked on the iNav codebase and don't currently have a lot of time free due to other commitments but if you get totally stuck let me know and perhaps I can help out.

**Comment by digitalentity** [comment]
@hydra Thanks for chiming in!

INAV's busDevice code initializes all defined CS# pins to logic "1" on boot, no problem with that. I'll consider the rest when reviewing this 😄

**Comment by hydra** [comment]
@digitalentity no worries!  Happy to help but I'm very time poor.  I corrected my previous comment, should have read motor outputs 5-8, not 4-5.

**Comment by digitalentity** [comment]
@hydra pretty much same here - time is a scarce resource. Cheers 🍻

**Comment by sdirkwinkel** [comment]
Thank you both for replying so quickly 👍 
I've added a commit to get both gyros cs pins initialized and use one of them. I've now got a working configurator  :raised_hands:
Getting a working debugger with openocd took some time but was very helpful.

I'll look further at gyro initialization and points 2/3. I'll also look through all defines in target.h to see which are actually used.

**Comment by hydra** [comment]
Great work.

I think also you might need to enable internal pullup for the gpio input for the sd card detect, i have a feeling i designed it without an external pull-up resistor.

**Comment by hydra** [comment]
@0xQSL Did you progress this any further?

Also, with regard to my previous comment regarding SD Card detect. I was wrong, there is no SD Card detect wired up to the MCU on the SP Racing F7 DUAL.  You just attempt to read the card at bootup and use it if it works.

**Comment by sdirkwinkel** [comment]
I've been using this on my hex with:

- 6x motor
- gps, compass (m8n)
- uarts for sbus and telemetry
- another uart for rf-module
- led strip
- sd-card (I'd checked your cleanflight code didn't see any cd-pin, so i just tried it and it worked)

I can't check vtx as i've got one of the stacking boards but need motors 5 and 6.
Alignment may also be wrong I just rotated it to match. Anything else works.

**Comment by hydra** [comment]
@digitalentity can we get this merged? We can add other hardware support on an as-needed basis.

The SPRacingF7DUAL is quite suited to iNAV due to its 8 motor output support, GPS connectors, on-board baro, extensive IO support and very capable CPU.

File: src/main/target/SPRACINGF7DUAL/target.h [hunk 1]
```diff
@@ -0,0 +1,232 @@
...
+
+#define USE_SERIAL_4WAY_BLHELI_INTERFACE
+
+#define TARGET_IO_PORTA         0xffff
+#define TARGET_IO_PORTB         0xffff
+#define TARGET_IO_PORTC         0xffff
+#define TARGET_IO_PORTD         (BIT(2))
+
+//#define USABLE_TIMER_CHANNEL_COUNT 16 // 4xPWM, 8xESC, 2xESC via UART3 RX/TX, 1xLED Strip, 1xIR.
+#define USABLE_TIMER_CHANNEL_COUNT 16
```
Comment (digitalentity) [inline src/main/target/SPRACINGF7DUAL/target.h]: This is no longer needed

File: src/main/target/SPRACINGF7DUAL/target.h [hunk 2]
```diff
@@ -0,0 +1,232 @@
...
+
+
+// unused but may be useful:
+// SPI3 is on the APB1 bus whose clock runs at 54MHz. Divide to under 400kHz for init:
+// #define SDCARD_SPI_INITIALIZATION_CLOCK_DIVIDER 256 // 54Mhz/256 = 210kHz
+// Divide to under 25MHz for normal operation:
+// #define SDCARD_SPI_FULL_SPEED_CLOCK_DIVIDER     4 // 54Mhz/3 = 18Mhz // FIXME currently running slower than 18mhz for testing.
+
+// #define SDCARD_DMA_STREAM_TX_FULL           DMA1_Stream7
+#define SDCARD_DMA_TX                       DMA1
```
Comment (digitalentity) [inline src/main/target/SPRACINGF7DUAL/target.h]: SDCARD definitions need refactoring. A lof of defines is not needed:

```
#define SPI3_CLOCK_LEADING_EDGE

#define USE_SDCARD
#define USE_SDCARD_SPI
#define SDCARD_SPI_BUS          BUS_SPI3
#define SDCARD_CS_PIN           PC3
```

File: src/main/target/SPRACINGF7DUAL/target.h [hunk 3]
```diff
@@ -0,0 +1,232 @@
...
+#define GYRO_MPU6500_ALIGN      GYRO_1_ALIGN
+
+#define USE_BARO
+#define BARO_I2C_BUS            BUS_I2C1
+#define USE_BARO_BMP280
+#define USE_BARO_MS5611
+
+#define MAG_I2C_BUS             BUS_I2C1
+#define USE_MAG
+#define USE_MAG_HMC5883
```
Comment (digitalentity) [inline src/main/target/SPRACINGF7DUAL/target.h:85]: Please add mode supported compasses, at least QMC5883

File: src/main/target/SPRACINGF7DUAL/target.c [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+#include "platform.h"
+#include "drivers/io.h"
+
+#include "drivers/timer.h"
+#include "drivers/timer_def.h"
+#include "drivers/dma.h"
+#include "drivers/bus.h"
+
+// Register second MPU6500 so cs pin is initialized
+BUSDEV_REGISTER_SPI(busdev_mpu6500_2,     DEVHW_MPU6500,      MPU6500_2_SPI_BUS,    MPU6500_2_CS_PIN,     GYRO_2_EXTI_PIN,  DEVFLAGS_NONE);
```
Comment (digitalentity) [inline src/main/target/SPRACINGF7DUAL/target.c:33]: It's better to decrare both gyros here and add `USE_TARGET_IMU_HARDWARE_DESCRIPTORS` definition to `target.h`

**Review by digitalentity — CHANGES_REQUESTED** [review]
Please address comments and rebase as this branch is no longer up to date with `development`. After that this could be merged.

**Comment by hydra** [comment]
@digitalentity thanks for review.   @0xQSL do you have time to make the requested changes and update the PR?

**Comment by sdirkwinkel** [comment]
Yes, I'll do it some time next week

**Comment by sdirkwinkel** [comment]
I've updated this pull request.
All sensors still read and motors still spin but i haven't flown with it yet.

**Comment by hydra** [comment]
@0xQSL Thanks for the update, any idea when will you get a change to test-fly it?

**Review by digitalentity — APPROVED** [review]

**Comment by sdirkwinkel** [comment]
> @0xQSL Thanks for the update, any idea when will you get a change to test-fly it?

I just did a quick test flight, everything works fine

**Comment by hydra** [comment]
@0xQSL Great!
@DigialEntity ready to merge now?

**Comment by digitalentity** [comment]
@0xQSL @hydra merged, thanks!

**Comment by hydra** [comment]
@digitalentity Great, thank you both too.

**Pull request merged**